### PR TITLE
chore: bug fix and optimization

### PR DESCRIPTION
**Pull request by miralandlabs**
Changes to atomicalsir:
1. sync fee estimation bug fix from atomicals-js to atomicalsir
2. add bitworkr/mint_bitworkr side-by-side with bitworkc/mint_bitworkc
3. set concurrency to num_cpus::get() - 1
4. fix max_fee comparison logic
5. add retry-send transaction control
6. log some values to assist tx recovery/replay in case that commit succeeds but reveal failed.
7. fine-tuning of reveal nonce generation mechanism to try to keep solution nonce as small as possible to save byte-storage.
8. add/change tracing log to improve user experience
9. add a command cli switch option -s to suppress detail tracing infos.

### CODE REVIEW

**Comment by AurevoirXavier** [comment]
Thanks for your PR.

1. I think we don't need the suppress flag. You can simply set that log level to debug or trace.
2. Any reason to set concurrency to `num_cpus::get() - 1`?

File: src/util.rs [hunk 1]
```diff
@@ -85,6 +85,7 @@ fn cbor_should_work() {
 		cbor(&PayloadWrapper {
 			args: Payload {
 				bitworkc: "aabbcc".into(),
+				bitworkr: Some(String::from("")),
```
Comment (AurevoirXavier) [inline src/util.rs:88]: Is this part necessary?

File: src/engine/rust.rs [hunk 2]
```diff
@@ -586,6 +754,7 @@ pub struct PayloadWrapper {
 #[derive(Debug, Serialize)]
 pub struct Payload {
 	pub bitworkc: String,
+	pub bitworkr: Option<String>,
```
Comment (AurevoirXavier) [inline src/engine/rust.rs:757]: Is this part necessary?

**Comment by miralandlabs** [comment]
> Thanks for your PR.
> 
> 1. I think we don't need the suppress flag. You can simply set that log level to debug or trace.
> 2. Any reason to set concurrency to `num_cpus::get() - 1`?

Yes, you are right. will remove suppress flag. As to num_cpus()::get() - 1, I originally didn't expect cpus running in full-capabilties, anyway, will follow your advices.

**Comment by AurevoirXavier** [comment]
> Yes, you are right. will remove suppress flag. As to num_cpus()::get() - 1, I originally didn't expect cpus running in full-capabilties, anyway, will follow your advices.

Then we can add a `thread` flag.

File: src/engine/rust.rs (hunk 2 above)
Comment (miralandlabs) [reply src/engine/rust.rs:757]: The intention is trying to keep sync and consistent with atomicals-js. Maybe not necessary.
I will submit a new PR soon for your review. Thanks for your advice.

**Comment by miralandlabs** [comment]
closed as some changes are made and submit a new PR

**Pull request closed without merging**